sklearn/utils/validation.py

```diff
@@ -402,13 +402,20 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
         array = np.array(array, dtype=dtype, order=order, copy=copy)
 
         if ensure_2d:
+            # If input is scalar raise error
+            if array.ndim == 0:
```

Comment:
Actually I think the error for scalar array should be the same irrespective of `ensure_2d`, i.e. it should be taken out of the `if ensure_2d` clause.

The error message when `ensure_2d==False` is not so great:
```py
from sklearn.utils import check_array
check_array(10, ensure_2d=False)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-ce2f2639811f> in <module>()
----> 1 check_array(10, ensure_2d=False)

~/dev/scikit-learn/sklearn/utils/validation.py in check_array(array, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    488     shape_repr = _shape_repr(array.shape)
    489     if ensure_min_samples > 0:
--> 490         n_samples = _num_samples(array)
    491         if n_samples < ensure_min_samples:
    492             raise ValueError("Found array with %d sample(s) (shape=%s) while a"

~/dev/scikit-learn/sklearn/utils/validation.py in _num_samples(x)
    118         if len(x.shape) == 0:
    119             raise TypeError("Singleton array %r cannot be considered"
--> 120                             " a valid collection." % x)
    121         return x.shape[0]
    122     else:

TypeError: Singleton array array(10) cannot be considered a valid collection.
```